### PR TITLE
cmake: Allow user to select bundled zlib

### PR DESCRIPTION
**Pull request by hkleynhans**
Under some circumstances the installed / system version of zlib may not
be desirable due to being too old or buggy.  This patch adds the option
`USE_BUNDLED_ZLIB` that will cause the bundled version of zlib to be
used.

We may also want to add similar functionality to allow the user to
select other bundled 3rd-party dependencies instead of using the system
versions.

/cc @pks-t @ethomson

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -255,19 +255,26 @@ ELSE()
...
+		ENDIF()
+		ADD_FEATURE_INFO(zlib ON "Zlib support")
+	ENDIF()
+ENDIF()
+IF (USE_BUNDLED_ZLIB OR NOT ZLIB_FOUND)
+	IF (USE_BUNDLED_ZLIB)
+	    MESSAGE(STATUS "zlib: user opted to use bundled 3rd-party sources." )
 	ELSE()
-		SET(LIBGIT2_PC_REQUIRES "${LIBGIT2_PC_REQUIRES} zlib")
+	    MESSAGE(STATUS "zlib was not found; using bundled 3rd-party sources." )
```
Comment (pks-t) [inline src/CMakeLists.txt]: Please use spaces here to indent that message

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -255,19 +255,26 @@ ELSE()
...
 
 # Optional external dependency: zlib
-FIND_PACKAGE(ZLIB)
-IF (ZLIB_FOUND)
-	LIST(APPEND LIBGIT2_INCLUDES ${ZLIB_INCLUDE_DIRS})
-	LIST(APPEND LIBGIT2_LIBS ${ZLIB_LIBRARIES})
-	IF(APPLE OR CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
-		LIST(APPEND LIBGIT2_LIBS "z")
-		LIST(APPEND LIBGIT2_PC_LIBS "-lz")
+IF (NOT USE_BUNDLED_ZLIB)
```
Comment (pks-t) [inline src/CMakeLists.txt]: Not a fault of yours, as we have just decided to enforce a consistent CMake style: please do not put any spaces between `IF(CONDITION)` instructions

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -255,19 +255,26 @@ ELSE()
...
+	FIND_PACKAGE(ZLIB)
+	IF (NOT USE_BUNDLED_ZLIP AND ZLIB_FOUND)
```
Comment (pks-t) [inline src/CMakeLists.txt]: `NOT USE_BUNDLED_ZLIB` is not required here, as it's already a prerequisite to getting here.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -255,19 +255,26 @@ ELSE()
...
+		LIST(APPEND LIBGIT2_INCLUDES ${ZLIB_INCLUDE_DIRS})
+		LIST(APPEND LIBGIT2_LIBS ${ZLIB_LIBRARIES})
+		IF(APPLE OR CMAKE_SYSTEM_NAME MATCHES "FreeBSD")
+			LIST(APPEND LIBGIT2_LIBS "z")
+			LIST(APPEND LIBGIT2_PC_LIBS "-lz")
+		ELSE()
+			SET(LIBGIT2_PC_REQUIRES "${LIBGIT2_PC_REQUIRES} zlib")
+		ENDIF()
+		ADD_FEATURE_INFO(zlib ON "Zlib support")
```
Comment (pks-t) [inline src/CMakeLists.txt]: Better: `ADD_FEATURE_INFO(zlib ON "using system zlib")`

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -255,19 +255,26 @@ ELSE()
...
+	ENDIF()
+ENDIF()
+IF (USE_BUNDLED_ZLIB OR NOT ZLIB_FOUND)
+	IF (USE_BUNDLED_ZLIB)
+	    MESSAGE(STATUS "zlib: user opted to use bundled 3rd-party sources." )
```
Comment (pks-t) [inline src/CMakeLists.txt]: I don't think this message is required. The user will probably know when he's opted in to that

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -255,19 +255,26 @@ ELSE()
...
 	ELSE()
-		SET(LIBGIT2_PC_REQUIRES "${LIBGIT2_PC_REQUIRES} zlib")
+	    MESSAGE(STATUS "zlib was not found; using bundled 3rd-party sources." )
 	ENDIF()
-	ADD_FEATURE_INFO(zlib ON "Zlib support")
-ELSE()
```
Comment (pks-t) [inline src/CMakeLists.txt]: Would be nice to have a `ADD_FEATURE_INFO(zlib ON "using bundled zlib")`

File: src/CMakeLists.txt (hunk 1 above)
Comment (hkleynhans) [reply src/CMakeLists.txt]: I think the rest of the file use tabs.  I have replaced by spaces with a tab which should look better.

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -255,23 +255,28 @@ ELSE()
...
+			LIST(APPEND LIBGIT2_PC_LIBS "-lz")
+		ELSE()
+			SET(LIBGIT2_PC_REQUIRES "${LIBGIT2_PC_REQUIRES} zlib")
+		ENDIF()
+		ADD_FEATURE_INFO(zlib ON "using system zlib")
+	ENDIF()
+ENDIF()
+IF(USE_BUNDLED_ZLIB OR NOT ZLIB_FOUND)
+	IF(NOT USE_BUNDLED_ZLIB)
+		MESSAGE(STATUS "zlib was not found; using bundled 3rd-party sources." )
```
Comment (pks-t) [inline src/CMakeLists.txt]: Sorry to be nitpicky here, didn't realize that in the first review .I'd love if this was actually part of the `IF(ZLIB_FOUND) ... ELSE()` branch, starting at line 270.

Comment (hkleynhans) [reply src/CMakeLists.txt]: Ah yes, that makes sense! 👍

**Comment by ethomson** [comment]
👍

**Pull request merged**